chalk-solve/src/clauses.rs

```diff
@@ -377,6 +384,57 @@ fn program_clauses_that_could_match<I: Interner>(
     Ok(clauses)
 }
 
+/// Adds the clause:
+///
+/// for<type, type> Normalize(<^0.0 as Trait>::Item -> ^0.1)
+///     :- Implemented(^0.0: Trait), Compatible, DownstreamType(^0.0), CannotProve
```

Comment:
Ok, so, I think this rule is right in spirit, but it won't work for more complex traits like

```rust
trait Foo<T> {
    type Bar;
}
```

Here, we would need something like

```
for<A, B, C> {
    Normalize(<A as Foo<B>>::Bar -> C) if
        Compatible, Implemented(A: Foo<B>), DownstreamType(A), CannotProve
}

for<A, B, C> {
    Normalize(<A as Foo<B>>::Bar -> C) if
        Compatible, Implemented(A: Foo<B>), FullyVisible(A), DownstreamType(B), CannotProve
}
```

Reply:
Compare to this logic

https://github.com/rust-lang/chalk/blob/b8a90d8122ec0589969c89b938777c8baf58be5b/chalk-solve/src/clauses/program_clauses.rs#L607-L623

Reply:
I think the latest commit fixes this.